### DISCUSSION TITLE
smallrye-pulsar cannot publish message in sample Quarkus project

### DISCUSSION THREAD

**Discussion by vszakd** [smallrye/smallrye-reactive-messaging#2567] [Q&A] [answered]
Hello,
I have a very simple demo project, made following the Quarkus quickstarts.

The project cannot publish because of this error:
```
WARN  [org.apa.pul.cli.imp.ProducerImpl] (vert.x-eventloop-thread-5) [quote-requests] [external-out] error while create opSendMsg by batch message container: java.lang.NoClassDefFoundError: com/scurrilous/circe/checksum/Crc32cIntChecksum
	at org.apache.pulsar.common.protocol.Commands.serializeCommandSendWithSize(Commands.java:1650)
	at org.apache.pulsar.common.protocol.Commands.newSend(Commands.java:568)
	at org.apache.pulsar.common.protocol.Commands.newSend(Commands.java:515)
	at org.apache.pulsar.client.impl.ProducerImpl.sendMessage(ProducerImpl.java:854)
	at org.apache.pulsar.client.impl.BatchMessageContainerImpl.createOpSendMsg(BatchMessageContainerImpl.java:255)
	at org.apache.pulsar.client.impl.ProducerImpl.batchMessageAndSend(ProducerImpl.java:2166)
	at org.apache.pulsar.client.impl.ProducerImpl.batchFlushTask(ProducerImpl.java:2149)
	at org.apache.pulsar.common.util.Runnables$CatchingAndLoggingRunnable.run(Runnables.java:54)
	at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:153)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:470)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:569)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.ClassNotFoundException: com.scurrilous.circe.checksum.Crc32cIntChecksum
	... 18 more
```

build.gradle
```
dependencies {
    implementation 'io.quarkus:quarkus-smallrye-reactive-messaging-pulsar'
    implementation enforcedPlatform("${quarkusPlatformGroupId}:${quarkusPlatformArtifactId}:${quarkusPlatformVersion}")
    implementation 'io.quarkus:quarkus-resteasy-reactive-jackson'
    implementation 'io.quarkus:quarkus-arc'
    testImplementation 'io.quarkus:quarkus-junit5'
    testImplementation 'io.rest-assured:rest-assured'
}
```

implementation
```
    @Outgoing("external-out")
    public Multi<Double> generate() {
        // Build an infinite stream of random prices
        // It emits a price every second
        return Multi.createFrom().ticks().every(Duration.ofSeconds(5))
                .map(x -> random.nextDouble());
    }
```

Thanks.

**Comment by ozangunalp** [comment]
Are you running it on JVM or native binary?

**Comment by ozangunalp** [comment]
Also a dependency tree would help, because that class is available in the `org.apache.bookkeeper:circe-checksum` module pulled by the `pulsar-client-original` dependency.

**Comment by vszakd** [comment]
I did further tests, it only happens with Gradle. The same identical
project with the same dependencies works fine on Maven.

I will send a dependency tree soon.

**Comment by vszakd** [comment]
JVM

**Comment by ozangunalp** [comment]
Interesting, they are all compile-time dependencies, I don't see how Gradle can resolve it differently.

**Comment by vszakd** [comment]
Just tried with two brand-new Maven and Gradle projects, created from the Quarkus "New project" wizard in IntelliJ IDEA using Quarkus 3.7.2.

Attached the two dependency trees.

[maven-dep.txt](https://github.com/smallrye/smallrye-reactive-messaging/files/14233630/maven-dep.txt)

[gradle-dep.txt](https://github.com/smallrye/smallrye-reactive-messaging/files/14233631/gradle-dep.txt)

**Comment by ozangunalp** [comment, marked as answer]
After much searching, I found out why this happens. It is not obvious but possible to resolve this.

The problem stems from the fact that the dependency artifact resolution in Maven and Gradle are different when it comes to the default _packaging_ of transitive dependencies. 

In Maven, the dependency type defaults to "jar". But in Gradle, a dependency is resolved by default to the type marked in the artifact description. For example, the `org.apache.bookkeeper:circe-checksum` artifacts packaging in pom file is `nar`, but there are artifacts published with type `nar`  `jar`, see https://repo1.maven.org/maven2/org/apache/bookkeeper/circe-checksum/4.16.1/.  
Therefore in Maven the transitive dependency is resolved to the jar type and included in the Quarkus app as expected. But in Gradle the `nar` type is resolved and filtered out from the dependencies. The same is true for native-image compilation with GraalVM.

So as to the resolution, I think the easiest way is to adjust the dependency resolution strategy in Gradle. You can either add explicit dependencies to the `jar` type artifacts: 
```
    implementation("org.apache.bookkeeper:circe-checksum:4.16.1") {
        artifact {
            type = "jar"
        }
    }
    implementation("org.apache.bookkeeper:cpu-affinity:4.16.1") {
        artifact {
            type = "jar"
        }
    }
```

or add dependency substitutions like 
```
configurations.all {    
    resolutionStrategy.dependencySubstitution {
        substitute(module("org.apache.bookkeeper:circe-checksum"))
            .using(module("org.apache.bookkeeper:circe-checksum:4.16.1"))
            .withoutClassifier()
    }
}
```


Even in Kotlin DSL I find advanced Gradle configurations are insanely complex so I'd suggest just declaring explicit dependencies.
Maybe you can find a better solution in Gradle [dependency resolutions rules](https://docs.gradle.org/current/userguide/resolution_rules.html#sec:substitution_with_attributes)

**Comment by cescoffier** [comment]
@ozangunalp should we explicitly declare that dependency in the Pulsar connector extension (runtime)?

**Comment by ozangunalp** [comment]
@cescoffier It won't change anything. The transitive dependency is already declared with jar type, default in Maven. Gradle doesn't resolve it as `jar`, but `nar`.

**Comment by vszakd** [comment]
@ozangunalp great find! Thanks.

**Comment by vszakd** [comment]
Should I open an issue with Quarkus to update the docs?

**Reply by cescoffier** [reply]
yes, please.

**Reply by vszakd** [reply]
Issue opened https://github.com/quarkusio/quarkus/issues/40628
